torch/cuda/__init__.py

```diff
@@ -317,41 +326,118 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
         raise AssertionError("Invalid peer device id")
     return torch._C._cuda_canDeviceAccessPeer(device, peer_device)
 
-
-@contextlib.contextmanager
-def stream(stream):
+class StreamContext(object):
     r"""Context-manager that selects a given stream.
-
     All CUDA kernels queued within its context will be enqueued on a selected
     stream.
-
-    Args:
-        stream (Stream): selected stream. This manager is a no-op if it's
+    Arguments:
+        StreamContext (Stream): selected stream. This manager is a no-op if it's
             ``None``.
-
     .. note:: Streams are per-device. If the selected stream is not on the
-        current device, this function will also change the current device to
-        match the stream.
+    current device, this function will also change the current device to
+    match the stream.
     """
-    if stream is None:
-        yield
-        return
-    src_prev_stream = current_stream()
-
-    if src_prev_stream.device != stream.device:
-        # The given stream is on a different device; have to restore the
-        # current_stream on that device on exit as well
-        with device(stream.device):
-            dst_prev_stream = current_stream()
-
-    torch._C._cuda_setStream(stream._cdata)
-    try:
-        yield
-    finally:
-        if src_prev_stream.device != stream.device:
-            torch._C._cuda_setStream(dst_prev_stream._cdata)
-        torch._C._cuda_setStream(src_prev_stream._cdata)
+    cur_stream : Optional['torch.classes.cuda.Stream']  # type: ignore
+    cur_stream_device: Optional[_device]
+
+    def __init__(self, stream: Optional['torch.classes.cuda.Stream']):  # type: ignore
+        self.idx = -1
+        self.stream = stream
+        self.prev_stream_device_index = -1
+        self.cur_stream_device_index = -1
+        # Initialize the below streams to default stream on the current device
+        self.device_index = _get_device_index(device=None, optional=True)
+        self.src_prev_stream = torch.cuda.default_stream(self.device_index)
+        self.dst_prev_stream = torch.cuda.default_stream(self.device_index)
 
+    def __enter__(self):
```

Comment:
The new context manager implementation is quite a bit longer than the old one, despite the eager API calling the same old functions in the end. Could there be more explanation about why it got longer like this?

Reply:
Part of the bloat comes from using a class with `__enter__` and `__exit__` defined instead of a function decorated with `@contextlib.contextmanager`. The former requires explicit initialization of the generator state from the latter, and JIT doesn't support the latter because it doesn't support generators (the decorated function needs to `yield` at some point).

Reply:
Even accounting for the loss of yield, the new code is still quite a lot longer (where it is querying the device state, where none of that happened in the old code)

Reply:
Here are the reasons as to why this grew longer was only because:
1. In JIT `@contextmanager` is not supported, so we had to implement by making a context manager class 
2. The eager mode `stream` contextmanager, uses `yield`, `try/finally` which are not supported in JIT. 
3. Many of the eager APIs (`torch._C._cuda_setDevice`, `current_device`,...) are not scriptable because few features of python are still not available(`try/finally`, `yield`,...) in Torchscript. Hence we have created similar API's in JIT which have same functionality but differ in the names.

Reply:
I think the part you're talking about can be removed - eager mode does that inside the implementation of `setStream`. If we do the same in JIT, we should be able to get rid of it. I commented it about it inline.

Reply:
I have now changed this and as suggested handled the setting of the device inside `set_stream`